### ISSUE TITLE
Allow skipping health check for a specific resource

### ISSUE THREAD

**Issue by aslafy-z** [argoproj/argo-cd#11782]
# Summary

Add a way to skip a health check on a specific resource / specific cluster.
We may think about allowing to select an health check overwrite per cluster instead of totally skipping the check.

# Motivation

Health checks implementation for Custom Resources are bound by the operator version. Here's one example:

I have an ArgoCD instance v2.5.5 that manages 2 clusters.
On cluster 1, prometheus-operator v0.58.0 set the status of the Prometheus resources.
On cluster 2, prometheus-operator v0.48.1 does not set any status for the Prometheus resources.
ArgoCD, in this version, includes a custom health check for the Prometheus resource, see https://github.com/argoproj/argo-cd/blob/v2.5.5/resource_customizations/monitoring.coreos.com/Prometheus/health.lua#L3. 

In order to disable the cluster 2 health check on the Prometheus resource, I had to overwrite the health check definition in argocd-cm like:

```lua
    resource.customizations.health.monitoring.coreos.com_Prometheus: |
      # BEGIN CUSTOMIZATION
      if obj.metadata ~= nil then
        if obj.metadata.annotations ~= nil then
          if obj.metadata.annotations["argocd.argoproj.io/skip-health-check"] ~= nil then
            hs = {}
            hs.status = "Healthy"
            hs.message = "Ignoring Prometheus Health Check"
            return hs
          end
        end
      end
      # END CUSTOMIZATION
      hs={ status = "Progressing", message = "Waiting for initialization" }
      if obj.status ~= nil then
        if obj.status.conditions ~= nil then
          for i, condition in ipairs(obj.status.conditions) do

            if condition.type == "Available" and condition.status ~= "True" then
              if condition.reason == "SomePodsNotReady" then
                hs.status = "Progressing"
              else
                hs.status = "Degraded"
              end
              hs.message = condition.message or condition.reason
            end
            if condition.type == "Available" and condition.status == "True" then
              hs.status = "Healthy"
              hs.message = "All instances are available"
            end
          end
        end
      end
      return hs
```

This issue was already reported in https://github.com/argoproj/argo-cd/issues/11261 and in #5620.

# Proposal

An easy impl would be to add a new annotation that asks the reconciler to skip the health assessment like `argocd.argoproj.io/skip-health-check`. This annotation can be added to any resources within Applications or even on Applications themselves.  

Allow overwriting the health assessment for all downstream resources in the Application manifests could also help to achieve more advanced use cases.

**Comment by pstatham-tx** [comment]
Rather than adding a prometheus annotation, what about something like this?

```lua
hs={ status = "Progressing", message = "Waiting for initialization" }
found_status = false 

if obj.status ~= nil then
  if obj.status.conditions ~= nil then
    for i, condition in ipairs(obj.status.conditions) do

      if condition.type == "Available" then
        found_status = true
        if condition.status ~= "True" then
            if condition.reason == "SomePodsNotReady" then
            hs.status = "Progressing"
            else
            hs.status = "Degraded"
            end
            hs.message = condition.message or condition.reason
        end
        if condition.status == "True" then
            hs.status = "Healthy"
            hs.message = "All instances are available"
        end
      end 
    end
  end
end

if not found_status then
    hs={ status = "Unknown", message = "Status is not provided" }
end

return hs
```

**Comment by leoluz** [comment]
> Rather than adding a prometheus annotation, what about something like this?

I also prefer the idea of enhancing the existing health check to make it work with different CRD versions. With more robust health check scripts, users wouldn't have to tweak the manifest to add annotation to skip the validation. However I am no opposed to add this additional annotation as well as an alternative.

**Comment by n89cz** [comment]
We would really appreciate this feature.

**Comment by Hronom** [comment]
Same here, I would like to have possibility to not mark application as degraded if one of it jobs failed

**Comment by TheBritz** [comment]
Interested in this

**Comment by crenshaw-dev** [comment]
I wouldn't be opposed to adding first-class support for `argocd.argoproj.io/skip-health-check`. Although it is technically possible to implement that today in resource customizations.

**Comment by Hronom** [comment]
> I wouldn't be opposed to adding first-class support for argocd.argoproj.io/skip-health-check. Although it is technically possible to implement that today in resource customizations.

@crenshaw-dev this option will work for me

**Comment by crenshaw-dev** [comment]
If anyone is willing to open a PR, I'd be happy to review!

**Comment by GabrielDyck** [comment]
Hi @crenshaw-dev ! What do you think about https://github.com/argoproj/argo-cd/pull/11901 ?

**Comment by danielloader** [comment]
This is quite the problem with crossplane providers, not all of them implement health statuses at all. It's quite a burden maintaining all of this in argocd where as I could put an annotation on the Provider object itself when they lack healthchecks.

This is in addition to https://docs.crossplane.io/knowledge-base/integrations/argo-cd-crossplane/#set-health-status for the ones that do behave...

**Comment by ShakataGaNai** [comment]
This is critical when dealing with 3rd party helm charts. Some of the tooling provided may not properly expose the health checks. While that's certainly not best practice (and getting those fixed is certainly the end goal), it's really problematic to have Argo think a deployment is forever ongoing because one small piece isn't healthy yet. 

When less educated users review the argo dashboard, they may think something is wrong when a deployment is still ongoing (and try to fix/rollback). Or worse, when someone is actively waiting for a deployment to show as finished... they'll be waiting a really REALLY long time.

**Comment by ebuildy** [comment]
Also another use case:

"Run Job or argo Workflow in background, without blocking application health"

Our specific use case: "run elasticsearch index that take 30 minutes, but not required for app health"

**Comment by chaehni** [comment]
> I wouldn't be opposed to adding first-class support for `argocd.argoproj.io/skip-health-check`. Although it is technically possible to implement that today in resource customizations.

How would I go about implementing this with customziation? I want to selectively enable health checks based on a resource label. Resources without the label should be skipped and not display any health info.

**Comment by paolostancato** [comment]
I would appreciate this feature too. In my case we have a Helm pre hook that pauses a ScaledObject, set replicas to 0 and runs a database migrations. Then a post hook un-pause the ScaledObject and everything is back to normal. With health checks the post hook never runs because Argo stops syncing the application

**Comment by kmadel** [comment]
I would like to see this for cluster status health checks as I don't believe there is a way to disable this. Or have a different `UserAgent` for the non-App cluster health check. I have vCluster instances connected to Argo CD and they auto-sleep and I would only like App syncs to wake the vCluster and ignore non-App sync health checks.

**Comment by aslafy-z** [comment]
I think this was implemented https://argo-cd.readthedocs.io/en/latest/operator-manual/health/#ignoring-child-resource-health-check-in-applications.

**Comment by apoole-q6cyber** [comment]
The healthcheck is ignored by the sync still fails. 
I'm using argocd version: `v2.14.11+8283115`

**Comment by jurgenweber** [comment]
yeah, same; I am trying to get a helm hook 'job' ignored, since it happens so fast ArgoCD doesn't notice and I Can not change the 

>ttlSecondsAfterFinished 

Setting.

I added 

> argocd.argoproj.io/ignore-healthcheck: "true"

into the job and pod spec, I confirm it is there when it flashes up, but ArgoCD is still syncing and will not continue

> waiting for completion of hook batch/Job/nginx-ingress-nginx-admission-create
